cirq/experiments/qubit_characterizations.py

```diff
@@ -0,0 +1,544 @@
+    rho_11 = numpy.mean(results.measurements['z'])
+    rho_00 = 1.0 - rho_11
+
+    circuit_x = circuit.copy()
+    circuit_x.append(ops.X(qubit) ** 0.5)
+    circuit_x.append(ops.measure(qubit, key='z'))
+    results = sampler.run(circuit_x, repetitions=num_shots)
+    rho_01_im = numpy.mean(results.measurements['z']) - 0.5
+
+    circuit_y = circuit.copy()
+    circuit_y.append(ops.Y(qubit) ** -0.5)
+    circuit_y.append(ops.measure(qubit, key='z'))
+    results = sampler.run(circuit_y, repetitions=num_shots)
+    rho_01_re = 0.5 - numpy.mean(results.measurements['z'])
+
+    rho_01 = rho_01_re + 1j * rho_01_im
+    rho_10 = numpy.conj(rho_01)
+
+    rho = numpy.array([[rho_00, rho_01], [rho_10, rho_11]])
+
+    return TomographyResult(rho)
+
+
+def two_qubit_state_tomography(sampler: sim.SimulatesSamples,
+                               q_0: devices.GridQubit, q_1: devices.GridQubit,
+                               circuit: circuits.Circuit,
+                               num_shots: int) -> TomographyResult:
+    """
+    Two-qubit state tomography.
+
+    To measure the density matrix of the output state of a two-qubit circuit,
+    nine z-basis measurements to obtain P_00, P_01 and P_10 are conducted,
+    preceded by different combinations of I, X/2 and Y/2 operations on the
+    two qubits. The results are store in a vector probs of length 3*9 = 27.
+    The density matrix rho is decomposed into an operator-sum representation
+    sum_{i, j} c_ij * numpy.kron(sigmas_i, sigmas_j), where i, j = 0, 1, 2,
+    3 and sigma_0 = I, sigma_1 = sigma_x, sigma_2 = sigma_y, sigma_3 =
+    sigma_z are the Identity and Pauli matrices.
+
+    Based on the measured probabilities probs and the transformations of the
+    measurement operator by different basis rotations, one can build an
+    overdetermined set of linear equations numpy.dot(mat, c) = probs. Here c
+    is of length 15 and contains all the c_ij's (except c_00 which is set to
+    1), and mat is a 27 by 15 matrix having three non-zero elements in each
+    row that are either 1 or -1.
+
+    The least-square solution to the above set of linear equations is used to
+    construct the density matrix rho.
+
+    See Vandersypen and Chuang, Rev. Mod. Phys. 76, 1037 for details.
+
+    Args:
+        sampler: The quantum engine or simulator to run the circuits.
+        q_0: The first qubit under test.
+        q_1: The second qubit under test.
+        circuit: The circuit to execute on the qubits before tomography.
+        num_shots: The number of measurements for each basis rotation.
+
+    Returns:
+        A StateTomographyResult object that stores and plots the density matrix.
+    """
+
+    def _measurement(two_qubit_circuit: circuits.Circuit) -> Sequence[float]:
+        two_qubit_circuit.append(ops.measure(q_0, q_1, key='z'))
+        results = sampler.run(two_qubit_circuit, repetitions=num_shots)
+        bit_strings = [r for r in results.measurements['z']]
```

Comment:
Use `results.histogram` instead of counting for yourself.

Reply:
Done. Cool, this really shortens things.